### PR TITLE
Fix: timeout field in subscriptions for mqtt and mqttCustom cases #4285

### PR DESCRIPTION
**Pull request by Anjali-NEC**
Fix issue #4285

### CODE REVIEW

File: src/lib/mqtt/MqttConnectionManager.cpp [hunk 1]
```diff
@@ -389,6 +417,11 @@ bool MqttConnectionManager::sendMqttNotification(const std::string& host, int po
     return false;
   }
 
+  if (timeoutInMilliseconds > 0)
+  {
+    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, timeoutInMilliseconds);
```
Comment (fgalan) [inline src/lib/mqtt/MqttConnectionManager.cpp:422]: This belong to the libcurl library, the one used to send HTTP notifications (MQTT notifications are sent using mosquitto library). However, note you are here in sendMqttNotification() function, so this doesn't make sense...

Comment (fgalan) [reply src/lib/mqtt/MqttConnectionManager.cpp]: Fixed in [a06ac67](https://github.com/telefonicaid/fiware-orion/pull/4313/commits/a06ac672070c368a6dddf831bedabae222076981)

**Comment by Anjali-NEC** [comment]
@fgalan sir,

Please review the PR and if it is ok then please merge it into master. Thanks

File: doc/manuals/orion-api.md [hunk 2]
```diff
@@ -3738,6 +3738,7 @@ A `mqtt` object contains the following subfields:
 | `qos`     | ✓        | number | MQTT QoS value to use in the notifications associated to the subscription (0, 1 or 2). If omitted then QoS 0 is used.                      |
 | `user`    | ✓        | string | User name used to authenticate the connection with the broker.                                                                             |
 | `passwd`  | ✓        | string | Passphrase for the broker authentication. It is always obfuscated when retrieving subscription information (e.g. `GET /v2/subscriptions`). |
+| `timeout` | ✓        | number | Maximum time (in milliseconds) the subscription waits for the response. The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-mqttTimeout` CLI parameter is used. See section in the [Command line options](admin/cli.md#command-line-options) for more details. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3741]:
```suggestion
| `timeout` | ✓        | number | Maximum MQTT broker connection time (in milliseconds). The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-mqttTimeout` CLI parameter is used. See section in the [Command line options](admin/cli.md#command-line-options) for more details. |
```

File: doc/manuals/orion-api.md [hunk 3]
```diff
@@ -3774,6 +3775,7 @@ A `mqttCustom` object contains the following subfields.
 | `payload` | ✓        | string | Text-based payload to be used in notifications. In case of empty string or omitted, the default payload (see [Notification Messages](#notification-messages) sections) is used. If `null`, notification will not include any payload. |
 | `json`    | ✓        | object | JSON-based payload to be used in notifications. See [JSON Payloads](#json-payloads) section for more details. |
 | `ngsi`    | ✓        | object | NGSI patching for payload to be used in notifications. See [NGSI payload patching](#ngsi-payload-patching) section for more details. |
+| `timeout` | ✓        | number | Maximum time (in milliseconds) the subscription waits for the response. The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-mqttTimeout` CLI parameter is used. See section in the [Command line options](admin/cli.md#command-line-options) for more details. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3778]:
```suggestion
| `timeout` | ✓        | number | Maximum MQTT broker connection time (in milliseconds). The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-mqttTimeout` CLI parameter is used. See section in the [Command line options](admin/cli.md#command-line-options) for more details. |
```

File: test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqtt.test
Comment (fgalan) [inline test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqtt.test:1]: Is this .test based in some existing test for HTTP timeout? In that case, could you provide name/link to that file, pls?

File: test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqttCustom.test [hunk 4]
```diff
@@ -0,0 +1,603 @@
+# Copyright 2023 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqttCustom.test:1]: Is this .test based in some existing test for HTTP timeout? In that case, could you provide name/link to that file, pls?

File: src/lib/mqtt/MqttConnectionManager.cpp [hunk 5]
```diff
@@ -43,6 +43,27 @@
 #define MQTT_DEFAULT_KEEPALIVE 60
 
 
+/* ****************************************************************************
+*
+* defaultTimeout -
+*/
+static long defaultTimeout = DEFAULT_TIMEOUT;
```
Comment (fgalan) [inline src/lib/mqtt/MqttConnectionManager.cpp:50]: Where this new variable is used?

(I see the new setMqttTimeout(), used in src/app/contextBroker/contextBroker.cpp, to set it, but I don't see in which places of the code it is actually used).

Comment (fgalan) [reply src/lib/mqtt/MqttConnectionManager.cpp:50]: In fact, we already have ` long timeout;` defined in src/lib/mqtt/MqttConnectionManager.h (L68). Is not a redundant setting? Why do we need this new one?

Comment (fgalan) [reply src/lib/mqtt/MqttConnectionManager.cpp:50]: Moreover, I don't see changes in `getConnection()` so it seems the new per-subscription timeout is no actually used in the MQTT broker connection logic.

**Comment by fgalan** [comment]
I have provided several feedback comments. Thanks for your contribution!

In addition to them, please solve the existing conflict in CHANGES_NEXT_RELEASE (it should be trivial: just apply your new line to the content existing in master branch) so functional test may pass.

File: src/lib/mqtt/MqttConnectionManager.cpp (hunk 5 above)
Comment (fgalan) [reply src/lib/mqtt/MqttConnectionManager.cpp:50]: Moreover bis :), taking into account that we have a pool of MQTT connections and that several subscription (each one with a possibly different `timeout`), what happend if two subscription share the same connection? The first one to be triggered is the one which `timeout` is taken into account.

Let me illustrate with an examples. We have these MQTT subscriptions:

* Sub A for entity EA, timeout=10, MQTT endpoint E
* Sub B for entity EB, timeout=20, MQTT endpoint E

So:

* If EA is update before EB, then sub A is triggered first, so timeout 10 would be used
* If EB is update before EA, then sub B is triggered first, so timeout 20 would be used

In sum: depending of which subscription is triggered first, one different timeout may be used, which is pretty confusing for the user.

**Comment by fgalan** [comment]
In fact, taking into account https://github.com/telefonicaid/fiware-orion/pull/4313#discussion_r1238137023, I'm not sure if this functionality makes sense... what do you think?

File: test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqtt.test
Comment (Anjali-NEC) [reply test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqtt.test:1]: I have taken the reference of below test case 
`test/functionalTest/cases/3842_per_sub_httpTimeout/per_sub_httpTimeout_crud.test`

File: test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqttCustom.test (hunk 4 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/4285_timeout_field_in_subscriptions_for_mqtt_and_mqttCustom_cases/timeout_field_in_subscriptions_for_mqttCustom.test:1]: `test/functionalTest/cases/3842_per_sub_httpTimeout/per_sub_httpTimeout_crud_custom.test`

**Comment by Anjali-NEC** [comment]
> In fact, taking into account [#4313 (comment)](https://github.com/telefonicaid/fiware-orion/pull/4313#discussion_r1238137023), I'm not sure if this functionality makes sense... what do you think?

I have the same understanding.

**Comment by fgalan** [comment]
> > In fact, taking into account [#4313 (comment)](https://github.com/telefonicaid/fiware-orion/pull/4313#discussion_r1238137023), I'm not sure if this functionality makes sense... what do you think?
> 
> I have the same understanding.

In that case, I suggest to close this PR and the associated issue, with an explanation on why mqttTimeout cannot be implemented in a per-subscription basic.

Could you proceed to close the PR, please? I will close the issue in sequence.

Thanks!

**Comment by Anjali-NEC** [comment]
> In that case, I suggest to close this PR and the associated issue, with an explanation on why mqttTimeout cannot be implemented in a per-subscription basic.
> 
> Could you proceed to close the PR, please? I will close the issue in sequence.

Closed PR#4313

**Pull request closed without merging**